File: Python/378__Kth_Smallest_Element_in_a_Sorted_Matrix.py
class Solution(object):
    def kthSmallest(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        def research(value):
            i, j = 0, n - 1
            number = 0
            while i <= n - 1 and j >= 0:
                if matrix[i][j] > value:
                    j -= 1
                if matrix[i][j] <= value:
                    i += 1
                    number += j+1
            return number

        n = len(matrix)
        left = matrix[0][0]
        right = matrix[-1][-1]
        while left < right:
            mid = (left + right) // 2
            mid_count = research(mid)
            if mid_count < k:
                left = mid + 1
            else:
                right = mid
        return left

File: Java/378__Kth_Smallest_Element_in_a_Sorted_Matrix.java
class Solution {
    public int kthSmallest(int[][] matrix, int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int size =  (matrix[0].length * matrix.length) - (k - 1);
        for (int[] row : matrix) {
            for (int i : row) {
                pq.add(i);
                if (pq.size() > size)
                    pq.poll();
            }
        }
        return pq.peek();

    }
}

// Binary Search
class Solution {
     public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        int left = matrix[0][0], right = matrix[n - 1][n - 1];
        while (left <= right) {
            int mid = (right + left) / 2;
            int num = count(matrix, mid);
            if (num >= k)
                right = mid - 1;
            else
                left = mid + 1;
        }
        return left;
    }

    private int count(int[][] matrix, int target) {
        int n = matrix.length;
        int result = 0;
        int i = n - 1, j = 0;
        while (i >= 0 && j < n) {
            if (matrix[i][j] <= target) {
                result += i + 1;
                j++;
            } else
                i--;
        }
        return result;
    }
}
